deep-learning/src/main/scala/com/microsoft/ml/spark/onnx/ONNXModel.scala

```diff
@@ -0,0 +1,496 @@
+        }
+
+        resultRow
+    }
+  }
+
+  private def createTensor(env: OrtEnvironment, tensorInfo: TensorInfo, batchedValues: Seq[_]) = {
+    val classTag = ClassTag(tensorInfo.`type`.clazz)
+    val flattened: Array[_] = flattenNested(batchedValues)(classTag).toArray
+
+    val shape: Array[Long] = tensorInfo.getShape
+    // the first dimension of the shape can be -1 when multiple inputs are allowed. Setting it to the real
+    // input size. Otherwise we cannot create the tensor from the 1D array buffer.
+    shape(0) = batchedValues.length
+
+    tensorInfo.`type` match {
+      case OnnxJavaType.FLOAT =>
+        val buffer = FloatBuffer.wrap(flattened.map(_.asInstanceOf[Float]))
+        OnnxTensor.createTensor(env, buffer, shape)
+      case OnnxJavaType.DOUBLE =>
+        val buffer = DoubleBuffer.wrap(flattened.map(_.asInstanceOf[Double]))
+        OnnxTensor.createTensor(env, buffer, shape)
+      case OnnxJavaType.INT8 =>
+        val buffer = ByteBuffer.wrap(flattened.map(_.asInstanceOf[Byte]))
+        OnnxTensor.createTensor(env, buffer, shape)
+      case OnnxJavaType.INT16 =>
+        val buffer = ShortBuffer.wrap(flattened.map(_.asInstanceOf[Short]))
+        OnnxTensor.createTensor(env, buffer, shape)
+      case OnnxJavaType.INT32 =>
+        val buffer = IntBuffer.wrap(flattened.map(_.asInstanceOf[Int]))
+        OnnxTensor.createTensor(env, buffer, shape)
+      case OnnxJavaType.INT64 =>
+        val buffer = LongBuffer.wrap(flattened.map(_.asInstanceOf[Long]))
+        OnnxTensor.createTensor(env, buffer, shape)
+      case OnnxJavaType.STRING =>
+        OnnxTensor.createTensor(env, flattened.map(_.asInstanceOf[String]), shape)
+      case other =>
+        throw new NotImplementedError(s"Tensor input type $other not supported. " +
+          s"Only FLOAT, DOUBLE, INT8, INT16, INT32, INT64, STRING types are supported.")
+    }
+  }
+
+  /**
+   * Returns true if the two data types are compatible. They are compatible if they share the same "shape", and
+   * 1. The element types from both sides are numeric types, or
+   * 2. The element types from both sides are the same.
+   */
+  @tailrec
+  private def compatible(from: DataType, to: DataType): Boolean = {
+    (from, to) match {
+      case (VectorType, right: ArrayType) =>
+        compatible(DoubleType, right.elementType)
+      case (left: ArrayType, right: ArrayType) =>
+        compatible(left.elementType, right.elementType)
+      case (_: NumericType, _: NumericType) => true
+      case (fromDataType, toDataType) => fromDataType == toDataType
+    }
+  }
+}
+
+class ONNXModel(override val uid: String)
+  extends Transformer
+    with ComplexParamsWritable
+    with ONNXModelParams
+    with Wrappable
+    with BasicLogging {
+
+  import ONNXModel._
+
+  override protected lazy val pyInternalWrapper = true
+
+  logClass()
+
+  def this() = this(Identifiable.randomUID("ONNXModel"))
+
+  @transient
+  lazy val modelInput: Map[String, NodeInfo] = {
```

Comment:
What exactly does this do? Should it be part of public API or private[ml] at least?

Reply:
This is meant to be a public API. The purpose is for users to see the ONNX input and output nodes (specifically, name, type and shape of these nodes), so users can map the feed/fetch dicts properly.

Reply:
i think this should be a def because you can change the model payload and this will not update if already accessed

Reply:
Fixed.